### ISSUE TITLE
Improving MediaPipe Lite Pose Detection in Fog and Low Light

### ISSUE THREAD

**Issue by ghost** [google-ai-edge/mediapipe#5996]
### Have I written custom code (as opposed to using a stock example script provided in MediaPipe)

None

### OS Platform and Distribution

Windows 10 Home

### Python Version

3.8.0

### Task name (e.g. Image classification, Gesture recognition etc.)

Pose Detection

### Describe the actual behavior

Unable to detect the person in fog

### Describe the expected behaviour

To be able to detect the person in fog

### Standalone code/steps you may have used to try to get what you need

I donwloaded a video from youtube - "The Grand Tour" where the guy is trying to look for his car in the fog, however even tho I can see the dude, mediapipe doesn't detect it. Only when the dude comes closer, it is able to detect it and create its pose landmarks. 

### Other info / Complete Logs

```shell
Additionally, there is a difference in behaviour observed when I resize the image using cv2
img = cv2.resize(img, (640, 480))
actually created a different BEHAVIOUR than
img = cv2.resize(img, (1300, 1000))
because the latter caused a car to be detected as a human pose!
```

**Comment by kuaashish** [comment]
Hi @PareekHeymun,

Could you please share the complete steps you are following from our documentation, or provide the full code snippet? This will help us better understand the issue.

Thank you!!

**Comment by ghost** [comment]
- I have dumped the code files in my github repo. Please find them. 
- I have added a readme.md for reference as well.
https://github.com/PareekHeymun/Mediapipe-Pose-Detection

**Comment by kuaashish** [comment]
Hi @PareekHeymun,

The codebase you shared uses the legacy Pose solution https://github.com/PareekHeymun/Mediapipe-Pose-Detection/blob/main/PoseModule.py#L15, which is no longer maintained or supported. This functionality has been integrated into the Pose Landmarker Task API.

We recommend migrating to the new Pose Landmarker API. You can find comprehensive documentation [here](https://ai.google.dev/edge/mediapipe/solutions/vision/pose_landmarker) and the Python implementation [here](https://ai.google.dev/edge/mediapipe/solutions/vision/pose_landmarker/python).

The new API offers [three models](https://ai.google.dev/edge/mediapipe/solutions/vision/pose_landmarker#models): Lite, Full, and Heavy. Each model has a corresponding model card detailing its limitations and other relevant information. We suggest trying the Full or Heavy models to improve person detection in your videos.

Thank you!!

**Comment by ghost** [comment]
Ok, thanks very much for the reply back, and thanks for providing clarity.
I do agree that I am using a 10-year-old laptop to run ML models, I need to buy a new one, or upgrade its parts.
Regardless, I am afraid that I cannot run the Full and Heavy models as my laptop hangs quite a lot.
After making necessary changes in my laptop, and trying out the FULL and HEAVY models, and if I find any issue, I will let you know. 
Thanks once again for your help.
I will close the issue now.

**Comment by ghost** [comment]
Ok @sumansuhag, I will find time and provide you with the details. Hence, I will reopen the issue. 
Yes, **downscaling and upscaling** as you said, are the key problems I noticed. Maybe I wasn't able to frame it well due to which @kuaashish wasn't able to get my point, but you have helped me out maybe, how to put it to the right terms :) 👍 
I will soon get back with more details.
Thank you.

**Comment by PareekHeymun** [comment]
MediaPipe Pose Detection Issues: Fog, Nighttime, and Resizing
Reported by: PareekHeymun
Issue: #5996
Date: June 13, 2025
To: @sumansuhaag, @kuaashish 

Setup:
MediaPipe: 0.10.11 (Lite model, model_complexity=0, min_detection_confidence=0.5)
Python: 3.8.0 (Conda conda3.8.0)
Dependencies: opencv-contrib-python==4.11.0.86, numpy==1.24.4, scikit-image==0.21.0
Code: [Mediapipe-Pose-Detection](https://github.com/PareekHeymun/Mediapipe-Pose-Detection) (Example Project.py, PoseModule.py)
Input: Videos (FOG_VISION.mp4 for fog, separate nighttime video)

Steps to Replicate

Clone [repository](https://github.com/PareekHeymun/Mediapipe-Pose-Detection).
Set up Conda environment:conda activate conda3.8.0
pip install mediapipe==0.10.11 opencv-contrib-python==4.11.0.86 scikit-image==0.21.0


Update video path in Example Project.py to FOG_VISION.mp4 (fog) or nighttime video.
Run: python Example_Project.py
Test resolutions: Modify cv2.resize to (640, 480) or (1300, 1000).

Issues

Fog:
Lite model fails to detect people in FOG_VISION.mp4 unless very close to the camera, due to low contrast and obscured landmarks.


Nighttime:
Detection works but is resolution-sensitive. At 640x480, distant subjects are missed; at 1300x1000, false positives occur (e.g., chair detected as human).


Resizing:
Downscaling (640x480) loses detail, reducing detection accuracy.
Upscaling (1300x1000) introduces artifacts, causing false positives.



Proposed Fixes

Model: Fine-tune Lite model for fog and low-light conditions.
Preprocessing: Add CLAHE and gamma adjustment to improve contrast in fog:img_float = img.astype(np.float32) / 255.0
img_lab = cv2.cvtColor(img_float, cv2.COLOR_BGR2LAB)
l, a, b = cv2.split(img_lab)
clahe = cv2.createCLAHE(clipLimit=2.0, tileGridSize=(8, 8))
l = clahe.apply((l * 255).astype(np.uint8))
img_lab = cv2.merge((l, a, b))
img = cv2.cvtColor(img_lab, cv2.COLOR_LAB2BGR)
img = exposure.adjust_gamma(img, gamma=0.8)


Resolution: Recommend optimal input size (e.g., 512x512) to match training data and reduce artifacts.
Hardware: Optimize Lite model for low-end devices, as Full/Heavy models are not feasible.

Notes

Tested preprocessing improved fog detection slightly but struggled with distant subjects.
512x512 resolution reduced some false positives vs. 1300x1000.
Increasing min_detection_confidence to 0.7 lowered false positives but reduced true positives.

Request
Please advise on model improvements or preprocessing integration for fog and resizing issues. Happy to test patches or provide logs/videos. Thanks for the support.

**Comment by ghost** [comment]
# Improving MediaPipe Lite Pose Detection in Fog and Low Light

**Reported by:** @PareekHeymun 
**Issue:** #5996
**Date:** June 14, 2025
**To:** @sumansuhag, @kuaashish

---

## Version Specifics

* **MediaPipe:** 0.10.11
* **Model:** Lite (`model_complexity=0`, `min_detection_confidence=0.5`, `min_tracking_confidence=0.5`)
* **Python:** 3.8.0 (Conda env `conda3.8.0`)
* **Dependencies:**
    * `opencv-contrib-python==4.11.0.86`
    * `numpy==1.24.4`
    * `scikit-image==0.21.0`

---

## Input

* Video (`FOG_VISION.mp4` for fog, separate nighttime video)

---

## Issues

* **Fog:** The Lite model fails to detect people in `FOG_VISION.mp4` unless they are very close to the camera, primarily due to low visibility.
* **Nighttime:** Detection at `640x480` misses distant subjects, while `1300x1000` causes false positives (e.g., a chair being detected as a human).
* **Resizing:** Downscaling to `640x480` results in a loss of detail, and upscaling to `1300x1000` introduces artifacts.

---

## Request

Can the **Lite model** be optimized for better performance in fog and low-light conditions? Alternatively, could you recommend an optimal input resolution (e.g., `512x512`) that would improve detection without exceeding hardware limitations? My current hardware restricts me to using only the Lite model. I'm happy to provide logs or assist with testing any potential fixes.

Thanks, @sumansuhag, @kuaashish!

**Comment by HarishPermal** [comment]
Hi, 

Sorry for the delayed response, the difficulty in detecting poses in fog and low light, as well as the varying behavior with  cv2.resize , stems from the inherent      
  challenges these conditions pose for computer vision models and the impact of input resolution on model performance. Models are typically trained  
  on clear images, and extreme conditions like fog introduce noise and reduce feature visibility, making detection difficult. Resizing images changes
  the input scale and can significantly alter the features the model perceives, sometimes leading to misclassifications if the resolution deviates   
  too much from the model's expected input (particularly if the aspect ratio gets distorted).                                                        
                                                                                                                                                     
  To potentially improve detection in challenging conditions, you can try applying image preprocessing techniques before feeding the frames to       
  MediaPipe. This might include contrast enhancement or denoising. For the  cv2.resize  issue, it's crucial to understand that the model's           
  performance is sensitive to input resolution. Experiment with different resolutions, ideally those close to the model's native input size, and     
  ensure aspect ratios are maintained or handled appropriately.                                                                                      
                                                                                                                                                     
  Here is a complete example of how you might apply some basic preprocessing (Contrast Limited Adaptive Histogram Equalization - CLAHE) to cut       
  through the fog, using the modern MediaPipe Tasks API:                                                                                             
                                                                                                                                                     
    import cv2                                                                                                                                       
    import mediapipe as mp                                                                                                                           
    from mediapipe.tasks import python                                                                                                               
    from mediapipe.tasks.python import vision                                                                                                        
    from mediapipe import solutions                                                                                                                  
    from mediapipe.framework.formats import landmark_pb2                                                                                             
    import numpy as np                                                                                                                               
                                                                                                                                                     
    # Note: You will need to download the model file first:                                                                                          
    # wget https://storage.googleapis.com/mediapipe-models/pose_landmarker/pose_landmarker_lite/float16/1/pose_landmarker_lite.task                  
                                                                                                                                                     
    def preprocess_frame(image):                                                                                                                     
        # Convert to grayscale for some operations                                                                                                   
        gray = cv2.cvtColor(image, cv2.COLOR_BGR2GRAY)                                                                                               
                                                                                                                                                     
        # Apply contrast limited adaptive histogram equalization (CLAHE)                                                                             
        clahe = cv2.createCLAHE(clipLimit=2.0, tileGridSize=(8,8))                                                                                   
        enhanced_gray = clahe.apply(gray)                                                                                                            
                                                                                                                                                     
        # Convert back to BGR for MediaPipe                                                                                                          
        enhanced_image = cv2.cvtColor(enhanced_gray, cv2.COLOR_GRAY2BGR)                                                                             
                                                                                                                                                     
        return enhanced_image                                                                                                                        
                                                                                                                                                     
    def draw_landmarks_on_image(rgb_image, detection_result):                                                                                        
        pose_landmarks_list = detection_result.pose_landmarks                                                                                        
        annotated_image = np.copy(rgb_image)                                                                                                         
                                                                                                                                                     
        # Loop through the detected poses to visualize.                                                                                              
        for idx in range(len(pose_landmarks_list)):                                                                                                  
            pose_landmarks = pose_landmarks_list[idx]                                                                                                
                                                                                                                                                     
            # Draw the pose landmarks.                                                                                                               
            pose_landmarks_proto = landmark_pb2.NormalizedLandmarkList()                                                                             
            pose_landmarks_proto.landmark.extend([                                                                                                   
                landmark_pb2.NormalizedLandmark(x=landmark.x, y=landmark.y, z=landmark.z) for landmark in pose_landmarks                             
            ])                                                                                                                                       
            solutions.drawing_utils.draw_landmarks(                                                                                                  
                annotated_image,                                                                                                                     
                pose_landmarks_proto,                                                                                                                
                solutions.pose.POSE_CONNECTIONS,                                                                                                     
                solutions.drawing_styles.get_default_pose_landmarks_style())                                                                         
        return annotated_image                                                                                                                       
                                                                                                                                                     
    # Initialize MediaPipe Pose Landmarker                                                                                                           
    base_options = python.BaseOptions(model_asset_path='pose_landmarker_lite.task')                                                                  
    options = vision.PoseLandmarkerOptions(                                                                                                          
        base_options=base_options,                                                                                                                   
        output_segmentation_masks=False)                                                                                                             
    detector = vision.PoseLandmarker.create_from_options(options)                                                                                    
                                                                                                                                                     
    # Example usage with a video frame                                                                                                               
    cap = cv2.VideoCapture('your_foggy_video.mp4')                                                                                                   
                                                                                                                                                     
    while cap.isOpened():                                                                                                                            
        ret, frame = cap.read()                                                                                                                      
        if not ret:                                                                                                                                  
            break                                                                                                                                    
                                                                                                                                                     
        # Resize the frame (experiment with optimal resolution, ensure aspect ratio is maintained)                                                   
        resized_frame = cv2.resize(frame, (640, 480))                                                                                                
                                                                                                                                                     
        # Apply preprocessing                                                                                                                        
        processed_frame = preprocess_frame(resized_frame)                                                                                            
                                                                                                                                                     
        # Convert the BGR image to RGB.                                                                                                              
        rgb_frame = cv2.cvtColor(processed_frame, cv2.COLOR_BGR2RGB)                                                                                 
                                                                                                                                                     
        # Create a MediaPipe Image object                                                                                                            
        mp_image = mp.Image(image_format=mp.ImageFormat.SRGB, data=rgb_frame)                                                                        
                                                                                                                                                     
        # Process the image and find poses.                                                                                                          
        detection_result = detector.detect(mp_image)                                                                                                 
                                                                                                                                                     
        # Draw the pose annotation on the image.                                                                                                     
        annotated_image = draw_landmarks_on_image(rgb_frame, detection_result)                                                                       
                                                                                                                                                     
        # Convert back to BGR for cv2 display                                                                                                        
        bgr_annotated = cv2.cvtColor(annotated_image, cv2.COLOR_RGB2BGR)                                                                             
                                                                                                                                                     
        cv2.imshow('MediaPipe Pose', bgr_annotated)                                                                                                  
        if cv2.waitKey(5) & 0xFF == 27:                                                                                                              
            break                                                                                                                                    
                                                                                                                                                     
    cap.release()                                                                                                                                    
    cv2.destroyAllWindows()                                                                                                                          
                                                                                                                                                     
  For more details on MediaPipe Pose Landmarker and its capabilities, please refer to the official documentation:                                    
  https://developers.google.com/mediapipe/solutions/vision/pose_landmarker                                                                           
                                                                                                                                                     
  Please try these suggestions and let us know if they improve detection for your use case.
